trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java

```diff
@@ -0,0 +1,337 @@
+
+        assertThatThrownBy(() -> readChunked(chunkedStream, TestingChunkSigningSession.build(GOOD_CREDENTIAL, BAD_SEED)))
+                .isInstanceOf(WebApplicationException.class);
+    }
+
+    @Test
+    public void testBadCredential()
+    {
+        TestingChunkSigningSession signingSession = TestingChunkSigningSession.build(GOOD_CREDENTIAL, GOOD_SEED);
+        String chunkedStream = signingSession.generateChunkedStream(GOOD_CONTENT, 3);
+
+        assertThatThrownBy(() -> readChunked(chunkedStream, TestingChunkSigningSession.build(BAD_CREDENTIAL, BAD_SEED)))
+                .isInstanceOf(WebApplicationException.class);
+    }
+
+    @Test
+    public void testMultipleExtensions()
+            throws IOException
+    {
+        TestingChunkSigningSession signingSession = TestingChunkSigningSession.build(GOOD_CREDENTIAL, GOOD_SEED);
+        String chunkedStream = signingSession.generateChunkedStream(GOOD_CONTENT, 3);
+        chunkedStream = chunkedStream.replace(";chunk-signature=", ";foo=bar;chunk-signature=");
+
+        assertThat(readChunked(chunkedStream, signingSession)).isEqualTo(GOOD_CONTENT.getBytes(UTF_8));
+    }
+
+    private byte[] readChunked(String chunkedStream, TestingChunkSigningSession signingSession)
+            throws IOException
+    {
+        try (InputStream in = new AwsChunkedInputStream(new ByteArrayInputStream(chunkedStream.getBytes(UTF_8)), signingSession)) {
+            return ByteStreams.toByteArray(in);
+        }
+    }
+
+    // NOTE: below vvvvvvv tests are modified from https://github.com/apache/httpcomponents-core/blob/e009a923eefe79cf3593efbb0c18a3525ae63669/httpcore5/src/test/java/org/apache/hc/core5/http/impl/io/TestChunkCoding.java
+
+    private static final String CHUNKED_INPUT
+            = "10;chunk-signature=0\r\n1234567890123456\r\n5;chunk-signature=0\r\n12345\r\n0;chunk-signature=0\r\n\r\n";
+
+    private static final String CHUNKED_RESULT
+            = "123456789012345612345";
+
+    @Test
+    public void testChunkedInputStreamLargeBuffer()
+            throws IOException
+    {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(CHUNKED_INPUT.getBytes(UTF_8));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        byte[] buffer = new byte[300];
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        int len;
+        while ((len = in.read(buffer)) > 0) {
+            out.write(buffer, 0, len);
+        }
+        assertEquals(-1, in.read(buffer));
+        assertEquals(-1, in.read(buffer));
+
+        in.close();
+
+        String result = out.toString(UTF_8);
+        assertEquals(CHUNKED_RESULT, result);
+    }
+
+    //Test for when buffer is smaller than chunk size.
+    @Test
+    public void testChunkedInputStreamSmallBuffer()
+            throws IOException
+    {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(CHUNKED_INPUT.getBytes(StandardCharsets.US_ASCII));
```

Comment:
Nit - should we use `UTF_8` here just like in the test above?

Reply:
Yeah. I meant to get all of those.